### DISCUSSION TITLE
The Future of Blitz

### DISCUSSION THREAD

**Discussion by ealmloff** [DioxusLabs/dioxus#1519] [General]
# The future of the Blitz renderer

Blitz is Dioxus' native renderer. Now that Servo is maintained, it is worth reconsidering the future of Blitz. Here are some possible options for what Blitz becomes in the future:

## Blitz becomes a browser renderer alternative to dioxus-desktop

Blitz was created as a browser renderer to replace a full webview in dioxus-desktop. We could continue to build out Blitz until it becomes a viable replacement.

Pros:
- No breaking changes
- We can optimize Blitz around templates to pre-calculate some of the cascading styles.
- Could potentially be more easily embeddable than existing browsers
- Fully written in modern rust which makes it easier to maintain than a chromium fork

Cons:
- Building a browser renderer is a monumental task
- Existing browsers may end up being more optimized because of how much time they have had to spend on optimization

## Blitz switches to use a browser renderer like chromium or Servo

We could use an existing browser renderer in Blitz like servo's incomplete `stylo` and `webrender` crates or we could fork chromium and create rust bindings.

Pros:
- No breaking changes
- Much more complete browser support
- Faster than a webview because we don't need to go through Javascript

Cons:
- We can't optimize around templates as easily
- The codebase is a lot larger and more difficult to maintain. It is either old rust code (`servo`) or c++ code (`chromium`).
- The same embed-ability as the underlying browser

**Comment by ealmloff** [comment]
Here are some other options that were discussed in this week's office hours:

## Dioxus adopts a different syntax from html more like react-native/Freya

We could adopt a syntax that is different from HTML to make native renderering more efficient (see react-native). Blitz would then render that syntax instead of HTML

Pros:
- Could potentially be more intuitive
- Potentially faster desktop rendering

Cons:
- Dioxus becomes much more difficult to learn coming from HTML or React
- We can't reuse `mdn`, `w3schools` and other web documentation to teach HTML. We need to create a whole new book with all of the options for styling, elements, and layout.
- Extremely breaking

## Blitz doesn't continue to exist in any form

We could abandon Blitz and just embrace the webview. Most desktop systems have a webview and browsers have full teams dedicated to maintaining and developing them. Performance isn't great because we need to continue to go through Javascript. A binary protocol in the short term and the wasm component model in the long term could make dioxus-desktop faster.

Pros:
- Much more complete browser support
- Much less code to maintain

Cons:
- We need to go through Javascript to change what the webview displays and handle events. This can bottleneck performance
- The same embed-ability as the underlying browser

**Comment by marc2332** [comment]
hi!

I am going to share my thoughts

I don't think `Blitz becomes a browser renderer alternative to dioxus-desktop` is worth, it takes a lot of development time while other solutions like Dioxus Desktop already just work

`Blitz switches to use a browser renderer like chromium or Servo` sounds interesting, but I would not go for Chromium, maintaining rusty chromium bindings is like an overkill idea, maybe servo is a better approach!

`Dioxus adopts a different syntax from html more like react-native/Freya` I think this is too breaking, and you would be adapting multiple renderers that already work fine just so Blitz could be better, I don't think it's worth. And I also think this would kind of "replace" Freya? I don't know how I feel about that ❤️

`Blitz doesn't continue to exist in any form` Not sure about this. I think it would be better to just have 1 desktop renderer as it simplifies development, but maybe it is still too early, perhaps it is still worth exploring Blitz

**Comment by ealmloff** [comment]
> Blitz switches to use a browser renderer like chromium or Servo sounds interesting, but I would not go for Chromium, maintaining rusty chromium bindings is like an overkill idea, maybe servo is a better approach!

The reason I mentioned Chromium is because it is a complete browser unlike Servo which is more complete than Blitz, but far from a full implementation of the spec.

> Dioxus adopts a different syntax from html more like react-native/Freya I think this is too breaking, and you would be adapting multiple renderers that already work fine just so Blitz could be better, I don't think it's not worth it. And I also think this would kind of "replace" Freya? I don't know how I feel about that ❤️

I agree! (To be clear these are just ideas transcribed from the office hours, not necessarily ideas I agree with). I think this is extremely unlikely. The benefits are unclear and it would break huge parts of the ecosystem and throw away a lot of what makes Dioxus easy to learn coming from React.

> Blitz doesn't continue to exist in any form Not sure about this. I think it would be better to just have 1 desktop renderer as it simplifies development, but maybe it is still too early, perhaps it is still worth exploring Blitz

It is possible that we get close to the performance of native bindings to the browser with a binary protocol. Latency could be an issue with a webview compared to Blitz

**Comment by lostb1t** [comment]
my 2 cents:

if you have the time and money, a native renderer would be technical the best solution.

However, devices have become more power-full and web engines have come a long way closing the gap. And yes being able to use html/css is a big plus for aome people.

My vote would be to keep webviews, but would indeed suggest investigating the option to have a webengine bundled for platforms that dont have webview support. (tvos, embedded?)  servo would be the obvious choice

out if scope: also some basic builtin components to close the gap to “native” would be helpfull. like viewtransitions and things like: https://tamagui.dev/docs/components/sheet/1.59.0

**Comment by mrgzi** [comment]
I came across this discussion while searching status of webrender in native platforms. First of all, I am not expert in this area but here are some of things coming to my mind.

- Dioxus can write some basic components (like in Flutter; Container, Text, Sizedbox etc..) which can be made with html and it can also render on Freya. When its compile to web, it can render regular html, for other platforms it can render on GPU using Freya. In this case Freya can merge into Dioxus repo.
- Servo probably is the best option. But is there any experiment using servo for rendering? (if servo is still lack of some features of web, then maybe Dioxus can just use implemented features for now?)
- @marc2332 Flutter developing another rendering engine called impeller because of jank animation issue. Do you think Freya also will have these kind of issue?
- Can it be possible use flutter engine for drawing some basic html and css elements? I think we don't need every html element, we need some basic html element for SEO. We can write our components with a few elements.

I would vote for flutter engine implementation because flutter already developing the engine so we will not have to maintain.

**Comment by jkelleyrtp** [comment]
I've been tinkering with Firefox's CSS engine for style resolution and things are looking decent.

https://github.com/jkelleyrtp/stylo-dioxus

The scope of rendering the web is large but tractable - and seemingly not so difficult to be impossible.

There's a lot more to the web than HTML and CSS, so we still need to think of how much we want to support long term.

**Comment by marc2332** [comment]
> * Dioxus can write some basic components (like in Flutter; Container, Text, Sizedbox etc..) which can be made with html and it can also render on Freya. When its compile to web, it can render regular html, for other platforms it can render on GPU using Freya. In this case Freya can merge into Dioxus repo.

This isn't exactly true, Freya doesn't use HTML, it's not an HTML/CSS renderer, it has it's own elements.

> @marc2332 Flutter developing another rendering engine called impeller because of jank animation issue. Do you think Freya also will have these kind of issue?

I haven't noticed any janks like Flutter has, and I think that is because they have some big shaders (something I haven't really used except for some demos in Freya) that are loaded dynamically at runtime and then cached after the first use, but that first load takes a little

**Comment by rjharmon** [comment]
... 17 months later, and considering Dioxus' and Servo's development since then, are there any updates on this?  

It's been enough time that some implicit understanding may have developed within the fabric of casual conversation between Dioxus developers and it would be great to load some fresh perspective on such an impactful topic near the center of Dioxus' "run anywhere" architecture.  

and, by the way, great work around here!  :racing_car: :tornado: :cloud:  - thanks!
